Web Apps: switched combobox to select2

orangejenny opened the pull request:
##### SUMMARY
https://dimagi-dev.atlassian.net/browse/USH-394

The atwho widget used for comboboxes is not at all accessible to screen readers. atwho is [no longer maintained](https://github.com/ichord/At.js/#%EF%B8%8F-announcement-%EF%B8%8F), so that's not going to improve.

select2's upcoming release has [a number of accessibility improvements](https://github.com/select2/select2/issues/3744), although it isn't yet released and there's [some debate](https://github.com/woocommerce/selectWoo/issues/35) about whether those improvements will be sufficient. My hope is that switching over to select2 will be enough of an improvement for the functionality needed in web apps, which is pretty basic. At a minimum, select2 is at least backed by a select element instead of a text box. We'll be testing this from an accessibility perspective and can consider additional changes, possibly switching to [selectWoo](https://developer.woocommerce.com/2017/08/08/selectwoo-an-accessible-replacement-for-select2/), if necessary.

This PR also updates "minimal" select questions, which were previously using plain HTML select elements, to use select2 for the sake of consistency. This makes the minimal and combobox options on web apps very similar, with the only difference being that the combobox allows for selecting the specific matching algorithm, whereas minimal uses select2's default search.

##### RISK ASSESSMENT / QA PLAN
[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-2058)

##### PRODUCT DESCRIPTION
Makes fairly minor changes to the appearance both both comboboxes and "minimal" select questions in web apps.

Before, comboboxes use atwho and minimal questions use standard HTML dropdowns:
![Screen Shot 2020-10-28 at 12 36 52 PM](https://user-images.githubusercontent.com/1486591/97467165-4c025f00-191a-11eb-9eb8-e53290f55bae.png)

![Screen Shot 2020-10-28 at 12 37 21 PM](https://user-images.githubusercontent.com/1486591/97467228-5c1a3e80-191a-11eb-9fc1-6c9f2abee9cb.png)

After, both widgets use select2:
![Screen Shot 2020-10-28 at 12 39 10 PM](https://user-images.githubusercontent.com/1486591/97467463-9c79bc80-191a-11eb-8f26-a65d014b8bab.png)


![Screen Shot 2020-10-28 at 12 24 31 PM](https://user-images.githubusercontent.com/1486591/97465634-9edb1700-1918-11eb-9a3e-118a4f5030c6.png)

orangejenny committed:
Fixed bindings for dropdown template

orangejenny committed:
Switched combobox to use dropdown template and select2 for rendering

orangejenny committed:
Added translations

orangejenny committed:
Combined DropdownEntry and ComboboxEntry to use the same UI

orangejenny committed:
Removed dropdown help text for consistency with prod

orangejenny commented:
FYI @dimagi/product

orangejenny committed:
Removed validation

No longer necessary because select2 doesn't allow you to enter free text.

biyeun commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js:
```diff
@@ -530,15 +530,39 @@ hqDefine("cloudcare/js/form_entry/entrycontrols_full", function () {
+            return "";
+        };
+
+        self.options = ko.computed(function () {
```
i'm guessing `pureComputed` doesn't have the `subscribe` option?

orangejenny replied:
It does, but the subscribers shouldn't have side effects. I think that rendering the select2 counts as a side effect, although I'm totally certain. I'm looking at the "When not to use a pure computed observable" section in these docs: https://knockoutjs.com/documentation/computed-pure.html

biyeun reviewed (approved):
this is looking great. is accessibility part of a USH initiative / is there some plan to looking into moving away from `atwho` on `vellum`? I'm sad another useful js library has now fallen into the unmaintained category. 😞

orangejenny commented:
@biyeun Agreed, I like atwho. Although some of the places we use it, like this one, the underlying concept is a dropdown and we only used atwho because the UI is lighter-weight than the old version of select2. In these places, I'm happy to replace with new select2. I think that applies to vellum's usages of atwho. Places where atwho is really being used as it's intended, like in the web apps debugger where it autocompletes function names in xpath expressions, are trickier.

As far as planning/resourcing, USH is indeed working on accessibility, but we're focused on web apps usage, so we're unlikely to update vellum anytime soon. Let's offline if you want to chat more about this.

orangejenny committed:
Added select2 to app preview template

orangejenny committed:
Updated to handle blank default value

orangejenny committed:
Extracted ComboboxEntry.filter again to make it available to tests

orangejenny committed:
Fixed lint error

orangejenny committed:
Added select2 to tests

orangejenny committed:
Restored onPreProcess and validation code

Even though the UI no longer allows invalid answers since it doesn't support free text,
it's still possible for another question to update a combobox with an invalid answer.

orangejenny committed:
Updated tests to reflect changes in entry's options

orangejenny committed:
Restored additional validation code

This validation is now available to both comboboxes and dropdowns.

biyeun reviewed (approved):
🥳

orangejenny committed:
Restored valueAllowUnset binding

Without this, web apps initially sets the value of an empty combobox to null,
but then the knockoout value binding sets it to undefined, which counts as a
change and triggers answer validation.

orangejenny committed:
Updated ComboboxEntry to use index for both answer and rawAnswer

This undoes 36ac455 and 0c19009, again removing the validation code
because dropdowns don't allow invalid selections. What those commits
missed is that the combobox needed to be updated to use the index for
rawAnswer. Previously, the combobox was backed by a plain text input,
which meant rawAnswer was text and the entry would look up that text's
index. Now that there's a dropdown, both the validation and the lookups
are no longer necessary. The only logic dealing with text is receiveMessage.

orangejenny committed:
Updated select2 so UI reflects programmatic value

orangejenny committed:
Updated dropdown and combobox to clear answer when options change

Previously, if options changed, the minimal dropdown would select whatever
option was at the same index as the previous answer, and the combobox
would should an 'invalid choice' error.

orangejenny commented:
@biyeun Requesting review once more, please - the last few commits changed validation logic a bit.

orangejenny committed:
Moved clearing logic to dropdown instead of combobox
